From 9f58c21e912d0a03cd1777403fe4a36b41c78653 Mon Sep 17 00:00:00 2001
From: Terry Lv <Terry.Lv@nxp.com>
Date: Fri, 8 Dec 2023 17:05:36 +0800
Subject: [PATCH 05/10] Add hr and spo2 support to pedometer driver.

Push pedometer driver with hr and spo2 support. The hr and spo2 reuse
pedometer framework.

Signed-off-by: Terry Lv <Terry.Lv@nxp.com>
---
 drivers/iio/imu/rpmsg_iio_pedometer.c | 218 ++++++++++++++++++++++----
 1 file changed, 190 insertions(+), 28 deletions(-)

diff --git a/drivers/iio/imu/rpmsg_iio_pedometer.c b/drivers/iio/imu/rpmsg_iio_pedometer.c
index 6edf347809a3..c201a7c37601 100644
--- a/drivers/iio/imu/rpmsg_iio_pedometer.c
+++ b/drivers/iio/imu/rpmsg_iio_pedometer.c
@@ -16,10 +16,21 @@
 #include <linux/module.h>
 
 
-#define RPMSG_IIO_DEV_STANDBY		0x00
-#define RPMSG_IIO_DEV_ACTIVED		0x01
-#define PEDOMETER_TYPE			0x00
-#define PEDOMETER_IDX			0x00
+#define RPMSG_IIO_DEV_STANDBY			0x00
+#define RPMSG_IIO_DEV_ACTIVED			0x01
+
+/*
+ * Note: as there's no IIO_HEARTRATE and IIO_SPO2 yet, we use IIO_COUNT and IIO_CONCENTRATION instead.
+ * */
+enum rpmsg_iio_sensor_type {
+	RPMSG_SENSOR_TYPE_PEDOMETER		= 0x00,
+	RPMSG_SENSOR_TYPE_TILT			= 0x01,
+	RPMSG_SENSOR_TYPE_HEARTRATE		= 0x02,
+	RPMSG_SENSOR_TYPE_SPO2			= 0x03,
+	RPMSG_SENSOR_TYPE_TEMP			= 0x04,
+	RPMSG_SENSOR_TYPE_TOTOL_NUM
+};
+#define RPMSG_SENSOR_IDX				0x00
 
 #define RPMSG_TIMEOUT_MS		1000
 
@@ -40,8 +51,8 @@ enum rpmsg_iio_header_type {
 };
 
 enum rpmsg_iio_header_cmd {
-	RPMSG_IIO_DETECTOR_CMD,
-	RPMSG_IIO_COUNTER_CMD,
+	RPMSG_IIO_STATE_DETECTOR_CMD,
+	RPMSG_IIO_DATA_REPORT_CMD,
 	RPMSG_IIO_POLL_DELAY_CMD,
 };
 
@@ -67,6 +78,13 @@ struct rpmsg_iio_data {
 	u32 steps;
 	u32 step_poll_delay;	// in ms
 
+	bool heartrate_enable;
+	bool spo2_enable;
+	bool temp_enable;
+	u32 heartrate_bps;
+	u32 spo2_rate;
+	s32 temp_val;
+
 	struct rpmsg_device *rpdev;
 	struct rpmsg_iio_msg *reply_msg;
 	struct rpmsg_iio_msg *notify_msg;
@@ -128,19 +146,58 @@ static inline void rpmsg_iio_event_report(struct rpmsg_iio_msg *msg)
 
 	indio_dev = iio_rpmsg->indio_dev;
 	ts = iio_get_time_ns(indio_dev);
-	if (msg->header.cmd == RPMSG_IIO_DETECTOR_CMD) {
+	switch (msg->header.cmd) {
+	case RPMSG_IIO_STATE_DETECTOR_CMD:
 		iio_push_event(indio_dev,
 			       IIO_EVENT_CODE(IIO_ACTIVITY, 0, IIO_NO_MOD,
 					      IIO_EV_DIR_RISING,
 					      IIO_EV_TYPE_THRESH, 0, 0, 0),
 			       ts);
-	} else if (msg->header.cmd == RPMSG_IIO_COUNTER_CMD) {
-		iio_rpmsg->steps = msg->instruct.val;
-		iio_push_event(indio_dev,
-			       IIO_EVENT_CODE(IIO_STEPS, 0, IIO_NO_MOD,
-					      IIO_EV_DIR_NONE,
-					      IIO_EV_TYPE_CHANGE, 0, 0, 0),
-			       ts);
+		break;
+	case RPMSG_IIO_DATA_REPORT_CMD:
+	{
+		switch (msg->sensor_type) {
+		case RPMSG_SENSOR_TYPE_PEDOMETER:
+			iio_rpmsg->steps = msg->instruct.val;
+			iio_push_event(indio_dev,
+							IIO_EVENT_CODE(IIO_STEPS, 0, IIO_NO_MOD,
+							IIO_EV_DIR_NONE,
+							IIO_EV_TYPE_CHANGE, 0, 0, 0),
+							ts);
+			break;
+		case RPMSG_SENSOR_TYPE_HEARTRATE:
+			iio_rpmsg->heartrate_bps = msg->instruct.val;
+			iio_push_event(indio_dev,
+							IIO_EVENT_CODE(IIO_COUNT, 0, IIO_NO_MOD,
+							IIO_EV_DIR_NONE,
+							IIO_EV_TYPE_THRESH, 0, 0, 0),
+							ts);
+			break;
+		case RPMSG_SENSOR_TYPE_SPO2:
+			iio_rpmsg->spo2_rate = msg->instruct.val;
+			iio_push_event(indio_dev,
+							IIO_EVENT_CODE(IIO_CONCENTRATION, 0, IIO_NO_MOD,
+							IIO_EV_DIR_NONE,
+							IIO_EV_TYPE_CHANGE, 0, 0, 0),
+							ts);
+			break;
+		case RPMSG_SENSOR_TYPE_TEMP:
+			iio_rpmsg->temp_val = msg->instruct.val;
+			iio_push_event(indio_dev,
+							IIO_EVENT_CODE(IIO_TEMP, 0, IIO_NO_MOD,
+							IIO_EV_DIR_NONE,
+							IIO_EV_TYPE_CHANGE, 0, 0, 0),
+							ts);
+			break;
+		default:
+			dev_err(&iio_rpmsg->rpdev->dev, "wrong rpmsg sensor type%d!\n",
+					msg->sensor_type);
+		}
+	}
+	break;
+	default:
+		dev_err(&iio_rpmsg->rpdev->dev, "wrong rpmsg command %d!\n",
+			msg->header.cmd);
 	}
 }
 
@@ -190,7 +247,7 @@ static struct rpmsg_driver rpmsg_iio_driver = {
 	.callback	= rpmsg_iio_cb,
 };
 
-static int rpmsg_iio_change_mode(int cmd, int enable)
+static int rpmsg_iio_change_mode(int cmd, int type, int index, int enable)
 {
 	struct rpmsg_iio_msg msg;
 
@@ -200,8 +257,8 @@ static int rpmsg_iio_change_mode(int cmd, int enable)
 	msg.header.minor = IMX_RMPSG_MINOR;
 	msg.header.type = RPMSG_IIO_SETUP;
 	msg.header.cmd = cmd;
-	msg.sensor_type = PEDOMETER_TYPE;
-	msg.sensor_index = PEDOMETER_IDX;
+	msg.sensor_type = type;
+	msg.sensor_index = index;
 	msg.instruct.inout.enable = enable;
 
 	return iio_send_message(&msg, iio_rpmsg);
@@ -217,8 +274,8 @@ static int rpmsg_iio_set_step_poll_delay(int cmd, int poll_delay)
 	msg.header.minor = IMX_RMPSG_MINOR;
 	msg.header.type = RPMSG_IIO_SETUP;
 	msg.header.cmd = cmd;
-	msg.sensor_type = PEDOMETER_TYPE;
-	msg.sensor_index = PEDOMETER_IDX;
+	msg.sensor_type = RPMSG_SENSOR_TYPE_PEDOMETER;
+	msg.sensor_index = RPMSG_SENSOR_IDX;
 	msg.instruct.val = poll_delay;
 
 	return iio_send_message(&msg, iio_rpmsg);
@@ -233,8 +290,26 @@ static int rpmsg_iio_read_event_value(struct iio_dev *indio_dev,
 {
 	switch (info) {
 	case IIO_EV_INFO_VALUE:
-		*val = iio_rpmsg->steps;
+	{
+		switch (chan->type) {
+		case IIO_STEPS:
+			*val = iio_rpmsg->steps;
+			break;
+		case IIO_COUNT:
+			*val = iio_rpmsg->heartrate_bps;
+			break;
+		case IIO_CONCENTRATION:
+			*val = iio_rpmsg->spo2_rate;
+			break;
+		case IIO_TEMP:
+			*val = iio_rpmsg->temp_val;
+			break;
+		default:
+			return -EINVAL;
+		}
 		return IIO_VAL_INT;
+	}
+	break;
 	case IIO_EV_INFO_PERIOD:
 		*val = iio_rpmsg->step_poll_delay;
 		return IIO_VAL_INT;
@@ -279,17 +354,31 @@ static int rpmsg_iio_read_raw(struct iio_dev *indio_dev,
 		switch (chan->type) {
 		case IIO_STEPS:
 			*val = pdata->counter_enable;
-			return IIO_VAL_INT;
+			break;
 		case IIO_ACTIVITY:
 			*val = pdata->detector_enable;
-			return IIO_VAL_INT;
+			break;
+		case IIO_COUNT:
+			*val = pdata->heartrate_enable;
+			break;
+		case IIO_CONCENTRATION:
+			*val = pdata->spo2_enable;
+			break;
+		case IIO_TEMP:
+			*val = pdata->temp_enable;
+			break;
 		default:
 			return -EINVAL;
 		}
+		break;
+	case IIO_CHAN_INFO_SCALE:
+		*val = 1000;  /* 62.5 */
+		*val2 = 16;
+		return IIO_VAL_FRACTIONAL;
 	default:
 		return -EINVAL;
 	}
-
+	return IIO_VAL_INT;
 }
 
 static int rpmsg_iio_write_raw(struct iio_dev *indio_dev,
@@ -306,25 +395,81 @@ static int rpmsg_iio_write_raw(struct iio_dev *indio_dev,
 		case IIO_STEPS:
 			mode = val ? RPMSG_IIO_DEV_ACTIVED:
 				     RPMSG_IIO_DEV_STANDBY;
-			ret = rpmsg_iio_change_mode(RPMSG_IIO_COUNTER_CMD, mode);
+			ret = rpmsg_iio_change_mode(RPMSG_IIO_STATE_DETECTOR_CMD,
+								RPMSG_SENSOR_TYPE_PEDOMETER, RPMSG_SENSOR_IDX, mode);
 			if (!ret)
 				pdata->counter_enable = val;
-			return ret;
+			break;
 		case IIO_ACTIVITY:
 			mode = val ? RPMSG_IIO_DEV_ACTIVED:
 				     RPMSG_IIO_DEV_STANDBY;
-			ret = rpmsg_iio_change_mode(RPMSG_IIO_DETECTOR_CMD, mode);
+			ret = rpmsg_iio_change_mode(RPMSG_IIO_DATA_REPORT_CMD,
+								RPMSG_SENSOR_TYPE_PEDOMETER, RPMSG_SENSOR_IDX, mode);
 			if (!ret)
 				pdata->detector_enable = val;
-			return ret;
+			break;
+		case IIO_COUNT:
+			mode = val ? RPMSG_IIO_DEV_ACTIVED:
+				     RPMSG_IIO_DEV_STANDBY;
+			ret = rpmsg_iio_change_mode(RPMSG_IIO_STATE_DETECTOR_CMD,
+								RPMSG_SENSOR_TYPE_HEARTRATE, RPMSG_SENSOR_IDX, mode);
+			if (!ret) {
+				ret = rpmsg_iio_change_mode(RPMSG_IIO_DATA_REPORT_CMD,
+								RPMSG_SENSOR_TYPE_HEARTRATE, RPMSG_SENSOR_IDX, mode);
+				if (!ret)
+					pdata->heartrate_enable = val;
+			}
+			break;
+		case IIO_CONCENTRATION:
+			mode = val ? RPMSG_IIO_DEV_ACTIVED:
+				     RPMSG_IIO_DEV_STANDBY;
+			ret = rpmsg_iio_change_mode(RPMSG_IIO_STATE_DETECTOR_CMD, RPMSG_SENSOR_TYPE_SPO2, RPMSG_SENSOR_IDX, mode);
+			if (!ret) {
+				ret = rpmsg_iio_change_mode(RPMSG_IIO_DATA_REPORT_CMD,
+								RPMSG_SENSOR_TYPE_SPO2, RPMSG_SENSOR_IDX, mode);
+				if (!ret)
+					pdata->spo2_enable = val;
+			}
+			break;
+		case IIO_TEMP:
+			mode = val ? RPMSG_IIO_DEV_ACTIVED:
+				     RPMSG_IIO_DEV_STANDBY;
+			ret = rpmsg_iio_change_mode(RPMSG_IIO_STATE_DETECTOR_CMD, RPMSG_SENSOR_TYPE_TEMP, RPMSG_SENSOR_IDX, mode);
+			if (!ret) {
+				ret = rpmsg_iio_change_mode(RPMSG_IIO_DATA_REPORT_CMD,
+								RPMSG_SENSOR_TYPE_TEMP, RPMSG_SENSOR_IDX, mode);
+				/* Do not need to set temp_enable here as temp is a self clear feature */
+			}
+			break;
 		default:
 			return -EINVAL;
 		}
+		break;
 	default:
 		return -EINVAL;
 	}
+
+	return ret;
 }
 
+static const struct iio_event_spec rpmsg_iio_heartrate_event = {
+	.type = IIO_EV_TYPE_CHANGE,
+	.dir = IIO_EV_DIR_NONE,
+	.mask_separate = BIT(IIO_EV_INFO_VALUE),
+};
+
+static const struct iio_event_spec rpmsg_iio_spo2_event = {
+	.type = IIO_EV_TYPE_CHANGE,
+	.dir = IIO_EV_DIR_NONE,
+	.mask_separate = BIT(IIO_EV_INFO_VALUE),
+};
+
+static const struct iio_event_spec rpmsg_iio_temp_event = {
+	.type = IIO_EV_TYPE_CHANGE,
+	.dir = IIO_EV_DIR_NONE,
+	.mask_separate = BIT(IIO_EV_INFO_VALUE),
+};
+
 static const struct iio_event_spec rpmsg_iio_step_event = {
 	.type = IIO_EV_TYPE_CHANGE,
 	.dir = IIO_EV_DIR_NONE,
@@ -338,11 +483,28 @@ static const struct iio_chan_spec rpmsg_iio_channels[] = {
 		.event_spec = &rpmsg_iio_step_event,
 		.num_event_specs = 1,
 	},
-
 	{
 		.type = IIO_ACTIVITY,
 		.info_mask_separate = BIT(IIO_CHAN_INFO_ENABLE),
 	},
+	{
+		.type = IIO_COUNT,
+		.info_mask_separate = BIT(IIO_CHAN_INFO_ENABLE),
+		.event_spec = &rpmsg_iio_heartrate_event,
+		.num_event_specs = 1,
+	},
+	{
+		.type = IIO_CONCENTRATION,
+		.info_mask_separate = BIT(IIO_CHAN_INFO_ENABLE),
+		.event_spec = &rpmsg_iio_spo2_event,
+		.num_event_specs = 1,
+	},
+	{
+		.type = IIO_TEMP,
+		.info_mask_separate = BIT(IIO_CHAN_INFO_ENABLE) | BIT(IIO_CHAN_INFO_SCALE),
+		.event_spec = &rpmsg_iio_temp_event,
+		.num_event_specs = 1,
+	}
 };
 
 static const struct iio_info rpmsg_iio_info = {
-- 
2.25.1

